### PR TITLE
[KernelGen][Nvidia] Add arctan2_ operator with Triton kernel

### PR DESCRIPTION
**Pull request by bwbwzzz**
## Summary
Adds a Triton kernel for `arctan2_`. Computes the element-wise arc tangent (inverse tangent) of `other/self(y/x)`, returning angles in radians between `-PI` and `PI`, and stores the result in-place in the input tensor.

## Testing
- Parametrized tests over 6 different shapes and float16/float32/bfloat16 dtypes (18 cases)
- Validated against PyTorch reference on device side
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_arctan2_.py -m arctan2_ -s --level core` (NVIDIA H20)

### float16
| Size Detail | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|---|---|---|---|
| [1073741824] | 4.370 | 4.308 | 1.014 |
| [64, 64] | 0.007 | 0.007 | 1.018 |
| [4096, 4096] | 0.075 | 0.076 | 0.991 |
| [64, 512, 512] | 0.075 | 0.076 | 0.988 |
| [1024, 1024, 1024] | 4.340 | 4.471 | 0.971 |

### float32
| Size Detail | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|---|---|---|---|
| [1073741824] | 9.394 | 9.538 | 0.985 |
| [64, 64] | 0.009 | 0.009 | 0.946 |
| [4096, 4096] | 0.156 | 0.156 | 1.000 |
| [64, 512, 512] | 0.157 | 0.155 | 1.011 |
| [1024, 1024, 1024] | 9.388 | 9.530 | 0.985 |

### bfloat16
| Size Detail | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|---|---|---|---|
| [1073741824] | 9.413 | 9.499 | 0.991 |
| [64, 64] | 0.009 | 0.009 | 1.022 |
| [4096, 4096] | 0.154 | 0.156 | 0.988 |
| [64, 512, 512] | 0.154 | 0.155 | 0.992 |
| [1024, 1024, 1024] | 19.186 | 9.501 | 2.019 |

**Arithmetic Mean Speedup: 1.00x**

## Multi-backend Testing
| Backend | Accuracy Test | Benchmark | Speedup (mean) | Notes |
|---|---|---|---|---|
| Nvidia (H20) | PASS | PASS (15 cases, --level core) | 1.00x | Primary |
| Tianshu | FAIL | PASS | 1.14x | — |
| Muxi | PASS | PASS | 1.01x | — |
| Ascend | PASS | — | — | Benchmark N/A |
| Hygon | PASS | PASS | 0.96x | — |

## Files Changed
- `src/flag_gems/ops/arctan2_.py`: Triton kernel implementation (in-place)
- `tests/test_arctan2_.py`: Accuracy test
- `benchmark/test_arctan2_.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: beta 5.1)

### CODE REVIEW

**Comment by bin913** [comment]
@bwbwzzz Please resolve conflicts

**Review by Dongxu-H — COMMENTED** [review]
## 🔍 Code Review

感谢您的贡献！这是一个实现良好的 in-place 操作符。我有一些问题和建议需要确认：

---

### ⚠️ 参数顺序需要确认

**当前代码：**
```python
def arctan2_(input, other):
    """In-place arctan2 operator: computes atan2(other, input) and stores result in input."""
    return arctan2_kernel(input, other, out0=input)
```

**PyTorch 语义：**
```python
# PyTorch: a.arctan2_(b) → 计算 atan2(a, b)，结果存入 a
# 等价于: a = torch.atan2(a, b)
```

**问题：**
- docstring 说 "computes atan2(other, input)"
- 但 kernel 调用是 `arctan2_kernel(input, other, ...)"
- 如果遵循 PyTorch 语义，应该是 `atan2(input, other)`

**建议：** 请确认参数顺序是否正确。可以添加一个简单的验证测试：
```python
# 验证参数顺序
x = torch.tensor([1.0, 0.0, -1.0], device='cuda')
y = torch.tensor([1.0, 1.0, 1.0], device='cuda')

# PyTorch reference
x_ref = x.clone()
x_ref.arctan2_(y)
# 应该等于 torch.atan2(x, y)

# Gems implementation
with flag_gems.use_gems():
    x_gems = x.clone()
    x_gems.arctan2_(y)

assert torch.allclose(x_ref, x_gems)
```

---

### 💡 测试覆盖建议

#### 1. 边界情况测试

当前测试使用随机数，缺少对边界值的验证。建议添加：

```python
@pytest.mark.arctan2_
def test_arctan2_edge_cases():
    # x = 0
    x = torch.tensor([0.0], device='cuda')
    y = torch.tensor([1.0, -1.0], device='cuda')
    # atan2(0, 1) = 0, atan2(0, -1) = π

    # y = 0
    x = torch.tensor([1.0, -1.0], device='cuda')
    y = torch.tensor([0.0, 0.0], device='cuda')
    # atan2(1, 0) = π/2, atan2(-1, 0) = -π/2

    # x = 0, y = 0
    x = torch.tensor([0.0], device='cuda')
    y = torch.tensor([0.0], device='cuda')
    # atan2(0, 0) = 0

    # Inf/NaN handling
    x = torch.tensor([float('inf'), float('-inf')], device='cuda')
    y = torch.tensor([1.0, 1.0], device='cuda')
```

#### 2. 广播测试

`arctan2_` 是 in-place 操作，不支持形状广播（因为结果必须存入 self）。但建议测试：
```python
# 验证广播错误处理
x = torch.randn(4, device='cuda')
y = torch.randn(2, 2, device='cuda')
with pytest.raises(RuntimeError):
    x.arctan2_(y)  # 应该报错：形状不匹配
```

---

### 📊 性能分析

**观察到的数据：**

| Dtype | Mean Speedup | Notable Case |
|---|---|---|
| float16 | ~1.00x | — |
| float32 | ~0.99x | — |
| bfloat16 | ~1.00x | [1024³]: **2.019x** |

**问题：**
1. 为什么 bfloat16 在大 tensor 上有显著加速（2.019x）？
   - PyTorch: 19.186ms
   - Gems: 9.501ms
   - 这看起来像是 PyTorch 在 bfloat16 大 tensor 上的性能问题

2. 整体加速比接近 1.0 是否符合预期？
   - 对于简单的 pointwise 操作，PyTorch 原生实现已经很快
   - Triton kernel 可能没有额外优化空间

---

### 🐛 Tianshu 后端测试失败

```
Tianshu: FAIL
```

**建议：** 
- 查看具体的错误日志
- 确认是否是数值精度问题还是其他问题
- 如果是已知问题，请在 PR 描述中注明

---

### ✅ 做得好的地方

1. **代码结构清晰** - 遵循项目模板，使用 `pointwise_dynamic` 装饰器
2. **输入验证** - 检查参数类型，提供清晰的错误信息
3. **PR 描述详细** - 包含测试、性能数据、多后端结果
4. **正确的注册** - operators.yaml, `__init__.py`, `_FULL_CONFIG` 都已更新
5. **使用 tl_extra_shim** - 正确处理 Triton 的数学函数

---

### 📋 后续步骤

**需要确认：**
1. ✅ 参数顺序是否正确（atan2(input, other) vs atan2(other, input)）
2. ✅ 多后端测试失败的具体原因

**建议改进：**
3. 添加边界情况测试（零值、无穷大、NaN）
4. 添加广播错误处理测试
5. 解决合并冲突（@bin913 提到）

---

整体来说这是一个质量很好的 PR！主要需要确认参数顺序和边界情况测试。期待您的回复！🙏

**Comment by bwbwzzz** [comment]
> @bwbwzzz Please resolve conflicts

have rosolved
